File: app01_java_nft/backend/src/main/java/com/gandarych/nftcerts/api/ArtistDashboardController.java
package com.gandarych.nftcerts.api;

import com.gandarych.nftcerts.certificate.Certificate;
import com.gandarych.nftcerts.certificate.CertificateService;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RestController;

import java.util.List;

/** GET /api/artists/{walletAddress}/dashboard — an artist's certificates and royalty overview. */
@RestController
public class ArtistDashboardController {

    private final CertificateService certificateService;
    private final CertificateDtoMapper certificateDtoMapper;

    public ArtistDashboardController(CertificateService certificateService, CertificateDtoMapper certificateDtoMapper) {
        this.certificateService = certificateService;
        this.certificateDtoMapper = certificateDtoMapper;
    }

    @GetMapping("/api/artists/{walletAddress}/dashboard")
    public ResponseEntity<ArtistDashboardResponse> dashboard(@PathVariable String walletAddress) {
        List<Certificate> certificates = certificateService.getCertificatesForWallet(walletAddress);
        List<CertificateDto> dtos = certificates.stream().map(certificateDtoMapper::toDto).toList();
        return ResponseEntity.ok(new ArtistDashboardResponse(walletAddress, dtos, dtos.size()));
    }
}

File: app02_cpp_nft/backend/src/api/ArtistDashboardController.cpp
#include "api/ArtistDashboardController.h"

#include <drogon/HttpAppFramework.h>

namespace nftcerts::api {

void registerArtistDashboardRoutes(certificate::CertificateService& certificateService,
                                    CertificateDtoMapper& certificateDtoMapper) {
    drogon::app().registerHandler(
        "/api/artists/{1}/dashboard",
        [&certificateService, &certificateDtoMapper](
            const drogon::HttpRequestPtr&, std::function<void(const drogon::HttpResponsePtr&)>&& callback,
            std::string walletAddress) {
            std::vector<certificate::Certificate> certificates =
                certificateService.getCertificatesForWallet(walletAddress);

            Json::Value certsJson(Json::arrayValue);
            for (const auto& cert : certificates) {
                certsJson.append(certificateDtoMapper.toDto(cert));
            }

            Json::Value response;
            response["walletAddress"] = walletAddress;
            response["certificates"] = certsJson;
            response["totalCertificates"] = static_cast<int>(certificates.size());

            callback(drogon::HttpResponse::newHttpJsonResponse(response));
        },
        {drogon::Get});
}

}  // namespace nftcerts::api
